### PR TITLE
Stateless models tree: Fix cases when visibility is determined incorrectly when category visibility is changed via selector.

### LINKED ISSUE #913: Tree widget: Models Tree category visibility determined incorrectly in specific case
http://localhost:3000/?iTwinId=892aa2c9-5be8-4865-9f37-7d4c7e75ebbf&iModelId=f34cc831-7139-4419-b502-b165e4294ca3

https://bentley.sharepoint.com/:v:/s/Rules-DrivenPresentation/Eci4fI6ADNdBkbQYV8i8Lz0B9Jg9PJ5ZOOIHRQT_SmHzLw?e=FwYHNC

### PR DESCRIPTION
**Pull request by Yato333**
Closes #913

### CODE REVIEW

File: change/@itwin-tree-widget-react-664e7245-3c8f-49a1-8d14-455f827301f5.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "none",
+  "comment": "Fix the way visibility is determined when category visibility is changed via selector",
+  "packageName": "@itwin/tree-widget-react",
+  "email": "yato333@users.noreply.github.com",
+  "dependentChangeType": "none"
```
Comment (grigasp) [inline change/@itwin-tree-widget-react-664e7245-3c8f-49a1-8d14-455f827301f5.json:6]:
```suggestion
  "type": "patch",
  "comment": "Fix the way visibility is determined when category visibility is changed via selector",
  "packageName": "@itwin/tree-widget-react",
  "email": "yato333@users.noreply.github.com",
  "dependentChangeType": "patch"
```

File: packages/itwin/tree-widget/src/test/trees/stateless/models-tree/internal/ModelsTreeVisibilityHandler.test.ts [hunk 2]
```diff
@@ -2252,6 +2252,165 @@ describe("HierarchyBasedVisibilityHandler", () => {
       });
     });
 
+    describe("interference from the category selector when there's no per model override", () => {
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/test/trees/stateless/models-tree/internal/ModelsTreeVisibilityHandler.test.ts:2255]:
```suggestion
    describe("reacting to category selector", () => {
```

File: packages/itwin/tree-widget/src/test/trees/stateless/models-tree/internal/ModelsTreeVisibilityHandler.test.ts [hunk 3]
```diff
@@ -2252,6 +2252,165 @@ describe("HierarchyBasedVisibilityHandler", () => {
...
+          viewport.renderFrame();
+
+          await validateHierarchyVisibility({
+            handler,
+            provider,
+            viewport,
+            visibilityExpectations: VisibilityExpectations.all("visible"),
+          });
+
+          viewport.changeCategoryDisplay([categoryId], true, true);
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/test/trees/stateless/models-tree/internal/ModelsTreeVisibilityHandler.test.ts:2401]: shouldn't this turn off the category?

File: packages/itwin/tree-widget/src/components/trees/stateless/models-tree/internal/ModelsTreeVisibilityHandler.ts [hunk 4]
```diff
@@ -343,7 +345,12 @@ class ModelsTreeVisibilityHandlerImpl implements ModelsTreeVisibilityHandler {
               elementsInBothAlwaysAndNeverDrawn: "model.elementsInAlwaysAndNeverDrawnList",
               noElementsInExclusiveAlwaysDrawnList: "model.noElementsInExclusiveAlwaysDrawnList",
             },
-            defaultStatus: createStatusByCategories,
+            defaultStatus: () => {
+              return createVisibilityStatus(
+                visibilityByCategories,
+                visibilityByCategories === "visible" ? "model.allCategoriesVisible" : "model.allCategoriesHidden",
+              );
+            },
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/components/trees/stateless/models-tree/internal/ModelsTreeVisibilityHandler.ts:353]:
```suggestion
            defaultStatus: () =>
              createVisibilityStatus(
                visibilityByCategories,
                visibilityByCategories === "visible" ? "model.allCategoriesVisible" : "model.allCategoriesHidden"
              ),
```

File: packages/itwin/tree-widget/src/components/trees/stateless/models-tree/internal/ModelsTreeVisibilityHandler.ts [hunk 5]
```diff
@@ -314,7 +314,16 @@ class ModelsTreeVisibilityHandlerImpl implements ModelsTreeVisibilityHandler {
...
-        map((categoryId) => this.getDefaultCategoryVisibilityStatus(categoryId, modelId)),
+        mergeMap((categoryId) => {
+          if (
+            viewport.perModelCategoryVisibility.getOverride(modelId, categoryId) === PerModelCategoryVisibility.Override.None &&
+            !viewport.view.viewsCategory(categoryId)
+          ) {
+            // If all category elements are always drawn it's actually visible.
+            return this.getCategoryDisplayStatus({ modelId, categoryId });
+          }
+          return of(this.getDefaultCategoryVisibilityStatus({ categoryId, modelId }));
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/components/trees/stateless/models-tree/internal/ModelsTreeVisibilityHandler.ts:325]: I don't understand why there's a special case for when the category is hidden through selector and there's no override. Can you explain?

File: packages/itwin/tree-widget/src/test/trees/stateless/models-tree/internal/ModelsTreeVisibilityHandler.test.ts (hunk 3 above)
Comment (Yato333) [reply packages/itwin/tree-widget/src/test/trees/stateless/models-tree/internal/ModelsTreeVisibilityHandler.test.ts]: Changed tests

**Review by grigasp — APPROVED** [review]

**Pull request merged**